### ISSUE TITLE
pickling the output of emcee.EnsembleSampler

### ISSUE THREAD

**Issue by pilchat** [dfm/emcee#229]
Hi,

I am running a simple code both in my machine and in a cluster. At the end of the code, I want to save the output of emcee.EnsembleSampler using pickle.dump.

In my machine(8 cores) everything goes fine. In the cluster, pickle.dump complains because:

TypeError: can't pickle mpi4py.MPI.Intracomm objects

In my PC and in the cluster I am using the same versions of emcee (2.2.1) and of mpi4py (1.3.1). Is it possible that emcee implicitly uses mpi4py2.0.0 (which is available in the cluster, not in my PC)? How can I check this?

In alternative, I can use mpi4py2.0.0, but then pickle.dump complains in my PC too. What do you suggest to save the output of emcee.EnsembleSampler and make it loadable later on?

Thanks in advance

**Comment by dfm** [comment]
I don't think that mpi pools are ever picklable. In the upcoming version of emcee, this should work by just deleting the pool before pickling.

**Comment by pilchat** [comment]
Hi Dan,

thanks for your reply. I've been trying to understand the reason for the issue, and my feeling is still that it depends on the version of the libraries and/or packages loaded by my code.

I confirm that I can pickle the emcee.EnsembleSampler (with the pool not closed yet) with emcee 2.2.1 and mpi4py 1.3.1 (inside ubuntu xenial).

Cheers

**Comment by dfm** [comment]
Interesting. Well I would recommend just adding:

```
del sampler.pool
```

before pickling to see if that works on the cluster.

If mpi4py 2.0 doesn't support pickling then that would be something to raise with them!
